streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
+            assignPendingStandbyTasksToLeastLoadedClients(clientsByUuid,
+                numStandbyReplicas,
+                standbyTaskClientsByTaskLoad,
+                tasksToRemainingStandbys);
+        }
+
+        return kafkaStreamsAssignments;
+    }
+
+    private static Map<ProcessId, KafkaStreamsAssignment> loadBasedStandbyTaskAssignment(final ApplicationState applicationState,
+                                                                                         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments) {
+        final int numStandbyReplicas = applicationState.assignmentConfigs().numStandbyReplicas();
+        final Map<TaskId, Integer> tasksToRemainingStandbys = applicationState.allTasks().values().stream()
+            .collect(Collectors.toMap(TaskInfo::id, taskInfo -> numStandbyReplicas));
+        final Map<ProcessId, KafkaStreamsState> streamStates = applicationState.kafkaStreamsStates(false);
+
+        final Set<TaskId> statefulTaskIds = applicationState.allTasks().values().stream()
+            .filter(TaskInfo::isStateful)
+            .map(TaskInfo::id)
+            .collect(Collectors.toSet());
+        final Map<UUID, KafkaStreamsAssignment> clients = kafkaStreamsAssignments.entrySet().stream().collect(Collectors.toMap(
+            entry -> entry.getKey().id(),
+            Map.Entry::getValue
+        ));
+
+        final ConstrainedPrioritySet standbyTaskClientsByTaskLoad = standbyTaskPriorityListByLoad(streamStates, kafkaStreamsAssignments);
+        standbyTaskClientsByTaskLoad.offerAll(streamStates.keySet().stream().map(ProcessId::id).collect(Collectors.toSet()));
+        for (final TaskId task : statefulTaskIds) {
+            assignStandbyTasksForActiveTask(numStandbyReplicas, clients,
+                tasksToRemainingStandbys, standbyTaskClientsByTaskLoad, task);
+        }
+        return kafkaStreamsAssignments;
+    }
+
+    private static void assignStandbyTasksForActiveTask(final int numStandbyReplicas,
+                                                        final Map<UUID, KafkaStreamsAssignment> clients,
+                                                        final Map<TaskId, Integer> tasksToRemainingStandbys,
+                                                        final ConstrainedPrioritySet standbyTaskClientsByTaskLoad,
+                                                        final TaskId activeTaskId) {
+        int numRemainingStandbys = tasksToRemainingStandbys.get(activeTaskId);
+        while (numRemainingStandbys > 0) {
+            final UUID client = standbyTaskClientsByTaskLoad.poll(activeTaskId);
+            if (client == null) {
+                break;
+            }
+            clients.get(client).assignTask(new AssignedTask(activeTaskId, AssignedTask.Type.STANDBY));
+            numRemainingStandbys--;
+            standbyTaskClientsByTaskLoad.offer(client);
+            tasksToRemainingStandbys.put(activeTaskId, numRemainingStandbys);
```

Comment:
nit: kind of confusing to update the map on every iteration of this loop, we should just do it once, when we have the final value after exiting the loop

(I also can't tell if the `tasksToRemainingStandbys` values are ever even accessed again after calling this map, but I'm not 100% sure and it seems best to leave that map in a consistent state just in case it does get used again in the future 🤷‍♀️ )